src/Routing.jl

```diff
@@ -0,0 +1,220 @@
+
+    all_basin_ids = read_routing_lv_basins(
+        joinpath(routing_lv_basins_dir, "all_basin_ids.txt"),
+    )
+    # make new txt file 
+    for basin_id in all_basin_ids # routing_lv_basins
+        timeseries_df =
+            CSV.read(joinpath(timeseries_dir, "basin_$basin_id.csv"), DataFrame)
+        filtered_df =
+            filter(row -> start_date <= row[:date] <= end_date, timeseries_df)
+        basin_area =
+            attributes_df[attributes_df.HYBAS_ID .== basin_id, :area][1]
+        runoff =
+            (filtered_df[:, :sro_sum] .+ filtered_df[:, :ssro_sum]) .*
+            basin_area ./ day_to_s .* km²_to_m²
+
+        streamflow = DSP.conv(runoff, distribution)[1:size(runoff)[1], :][:]
+
+        output_df = DataFrame(date = dates, streamflow = streamflow)
+
+        CSV.write(
+            joinpath(output_dir, "hillslope_basin_$basin_id.csv"),
+            output_df,
+        )
+    end
+end
+
+# Recursive function to get a list of all upstream basins
+function get_upstream_basins(basin_id::String, graph_dict::Dict{String, Any})
+    # Base case: if the current basin has no upstream basins, return an empty list
+    if isempty(graph_dict[basin_id])
+        return []
+    end
+
+    # Initialize the list of upstream basins
+    upstream_basin_list = []
+
+    # Iterate over the direct upstream basins of the current basin
+    for up_basin in graph_dict[basin_id]
+        # Add the current upstream basin
+        push!(upstream_basin_list, up_basin)
+
+        # Recursively collect upstream basins of this upstream basin
+        append!(
+            upstream_basin_list,
+            get_upstream_basins(string(up_basin), graph_dict),
+        )
+    end
+
+    # Remove duplicates to ensure each basin appears only once
+    return unique(upstream_basin_list)
+end
+
+function update_state_from_channel!(
+    river_state::RS,
+    channel_model::CM,
+    env::E,
+    start_date::Date,
+    end_date::Date,
+) where {RS <: RiverState, CM <: AbstractChannelModel, E <: Environment}
+    println("Starting channel update")
+
+    km_to_m = 1e3
+
+    output_dir = env.dyn_env.output_dir
+    timeseries_dir = env.dyn_env.timeseries_dir
+
+    routing_lv_basins_dir = env.static_env.routing_lv_basins_dir
+    attributes_df = CSV.read(
+        joinpath(env.static_env.attributes_dir, "attributes.csv"),
+        DataFrame,
+    )
+
+    C, D = channel_model.wave_velocity, channel_model.diffusivity
+    t_max = channel_model.t_max
+
+    all_basin_ids = read_routing_lv_basins(
+        joinpath(routing_lv_basins_dir, "all_basin_ids.txt"),
+    )
+
+    # Iterate over basins in the given routing level
+    for basin_id in all_basin_ids
+        # Get current streamflow in the basin
+        timeseries_df = CSV.read(
+            joinpath(output_dir, "hillslope_basin_$basin_id.csv"),
+            DataFrame,
+        )
+
+        # Allocate streamflow from upstreams
+        up_streamflow = zeros(length(timeseries_df[:, :streamflow]))
```

Comment:
Ideally all of these allocations should be done once before the simulation starts - should this be part of the state? but we can discuss!

Reply:
to discuss in meeting (12/6)
- need empty streamflow for each basin